### DISCUSSION TITLE
Performance Debugging with my application on Nanos

### DISCUSSION THREAD

**Discussion by Roopsai507** [nanovms/nanos#2042] [Q&A]
I've been testing a server application on both Linux and the Nanos Unikernel. The application binary starts a server and handles requests. On Linux, the total response time for a request is consistently around 100-150ms. However, when running the same application on Nanos, the response time increases significantly to around 1.6-2 seconds.

To investigate this issue, I enabled tracing and ran the application using the following command:
```
ops pkg load myapp --trace -p 8080 -c config.json --syscall-summary
```
The trace logs, however, did not provide clear insights into the cause of the latency. Here is a snippet from the logs:
```
3 thread_wakeup: 3->3 blocked_on thread, RIP=0x13351ce2e
        3 sleep interruptible (on "thread")
        3 thread_wakeup: 3->3 blocked_on thread, RIP=0x13351ce2e
    2 thread_wakeup: 2->2 blocked_on futex, RIP=0x133488117
        3 sleep interruptible (on "thread")
            4 sleep interruptible (on "futex")
            4 thread_wakeup: 4->4 blocked_on futex, RIP=0x1334882c0
    2 sleep interruptible (on "futex")
                    6 sleep interruptible (on "futex")
                    6 thread_wakeup: 6->6 blocked_on futex, RIP=0x1334882c0
        3 thread_wakeup: 3->3 blocked_on thread, RIP=0x13351ce2e
        3 sleep interruptible (on "thread")
                                        11 sleep interruptible (on "futex")
                                        11 thread_wakeup: 11->11 blocked_on futex, RIP=0x1334882c0
                                    10 sleep interruptible (on "futex")
                                    10 thread_wakeup: 10->10 blocked_on futex, RIP=0x1334882c0
                                    10 sleep interruptible (on "futex")
        3 thread_wakeup: 3->3 blocked_on thread, RIP=0x13351ce2e
        3 sleep interruptible (on "thread")
                                    10 thread_wakeup: 10->10 blocked_on futex, RIP=0x1334882c0
                    6 sleep interruptible (on "futex")
                    6 thread_wakeup: 6->6 blocked_on futex, RIP=0x1334882c0

```

The logs mainly indicate threads being woken up and blocked on futexes, but they do not pinpoint the source of the latency.
How can I effectively debug the increased response time when running on Nanos Unikernel? Are there specific tools or methods that can help identify the root cause of this slowness?

Any insights or suggestions would be greatly appreciated.

**Comment by eyberg** [comment]
That definitely sounds problematic. ftrace might come in handy here, https://nanovms.com/dev/tutorials/profiling-and-tracing-nanos-unikernel but to start with - how are you comparing these instances? Are they in the cloud? local? Are they both using kvm and are  bridged? Can you describe your setup?

Do you have a smaller bit of sample code that is reproducible that you can share?

**Comment by Roopsai507** [comment]
I am comparing the application performance by running it in a Firecracker microVM using a Debian 12 OS image, as well as running it with OPS, both in qemu and within Firecracker. Both setups showed higher latency compared to running the application directly on the Debian image. All testing is being conducted on my local machine, which is running Ubuntu 20.04.6 (Focal Fossa). Unfortunately, I do not have a sample code snippet available to reproduce the error.

I've created a package under ops/packages/ that includes the application binary and its necessary files(html,css..etc). To execute the application, I use the following command:
```
ops pkg load myapp --trace -p 8080 -c config.json --syscall-summary
```

Could you provide steps on how to enable Ftrace for further investigation?

**Comment by eyberg** [comment]
besides the linked article there is also this, https://github.com/nanovms/nanos/tree/master/tools/trace-utilities , you could also use something like perf, https://nanovms.com/dev/tutorials/measuring-unikernel-cold-boot-time-with-perf

but i kinda feel like they would both be overkill for what you're seeing

I don't know how helpful these tools are going to be for you here - if you're saying a request is taking 1-2 seconds when it should be taking 100-200ms i'd start by timestamping the beginning and end of that request in your request handler and use that as your starting point and then putting in more timestamps halfway, then halfway between that, etc. in between to drill down where it exactly you're seeing an issue

this doesn't sound like it's something nanos related, although it most certainly could be;  but it's really hard to know what you're running into without some kind of reproducible code that shows it

i'm not even sure what language you're using here but you could make a simple webserver in that language that either shows this behavior and we can go from there or it doesn't show that behavior and that gives you a clue that there's something inside that request handler that is causing issues; or you could even put a new simple hello route in that app and see if it does the same thing or not

**Comment by Roopsai507** [comment]
I’m looking to enable ftrace in the ops image. The command I’ve come across is:
```
make TRACE=ftrace TARGET=my-server run
```
I've gone through the ftrace tutorial however, I'm unclear on what the TARGET parameter specifically refers to in this context. 

Additionally, I need some steps to compile Nanos with my application(consisting of multiple folders/files) and enable ftrace.

**Reply by eyberg** [reply]
The target refers to runtimes tests that live in test/runtime - so for example 'webg' is listed in the Makefile as the target and it has code and a corresponding manifest file in test/runtime/webg.manifest that details all the stuff that would go on the fs but that's not really something that you would want to do by hand, especially for a fs that is large and that's one reason why we have ops auto-generate it.

This is really going to be more helpful if you are digging into nanos itself vs something in your app.

Have you tried my earlier suggestion of just comparing a simple request handler against the one that is giving you problems? It would be good to know if you are seeing the same issue with a "hello world" request/response as that would tell us something else is going on.

**Reply by Roopsai507** [reply]
While comparing the performance of a simple hello-world request/response, seemed to work fine. However, when more complex tasks like inserting and retrieving data through an endpoint are involved, I noticed an increase in latency.

I tried using the --syscall-summary flag with the ops run command to obtain a syscall summary similar to the output from strace -c. 
```ops run --syscall-summary appbinary -p 8080 -c config.json```
 the expected syscall summary didn’t get printed for the ELF binary.
 Could tell the steps to print this syscall summary

**Reply by francescolavra** [reply]
When enabled, the syscall summary is printed by the kernel just before shutting down; however, be aware that terminating a local instance running with `ops run` by just pressing CTRL-C doesn't do a clean shutdown, it just kills the underlying Qemu process without notifying the guest VM, so the guest kernel doesn't have a chance to print out the syscall stats. What you may want to do is exporting an endpoint that instructs your server application to terminate, so when you do a request to that endpoint the kernel in the server will print out the stats and then terminate.
Another thing to be aware of is that syscall tracing is not implemented within the kernel core code, but it's in a separate klib callled strace, so the `--syscall-summary` flag takes effect only if your image contains the strace klib. You can include this klib in your image by inserting the following in your config.json file:
```
"Klibs": ["strace"]
```
In a future Ops release the inclusion of this klib will be done automatically whenever the command line flags require it, so it won't be necessary to insert the above line in the config file.

**Reply by eyberg** [reply]
Ok - so it sounds like it is related to this specific request. That's good - now you can isolate that. Is the endpoint you are referring to a website or a database or some other service it's talking too? If it's something public than we could start trying to reproduce that. If it's not it could potentially be mocked. If you say that it is talking to a mysql db or it's talking to a java app - these are the types of details that would be good to know.

Now that you know that a request handler like this is causing issues you can use that binsearch method to further isolate the issue, however, it sounds like you already kinda feel that it is when it's talking to this endpoint.

I didn't ask before because you had mentioned you were using it in both firecracker && qemu but you can easily get different results using something like 'ops run' depending on your environment. Showing us the output of 'ops profile' would probably be helpful here.

What language are you working in? Being as specific as possible on all of this makes it easier to help.

Also, just to tack on to the prior comment you had originally asked about "ftrace" but the --syscall-summary is a part of the "strace" klib - these are different.

**Reply by Roopsai507** [reply]
About the application:
The application is a C-based server that uses an embedded RocksDB database, compiled with clang. It includes a /get endpoint that retrieves all users previously inserted into the database. When a request is made to this endpoint, the server fetches and returns the stored user data.



I have added a /exit endpoint and used strace --syscall-summary to check for any syscall latency. 

```strace -c ./my_app```

```
time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 98.67    0.022402        1317        17         1 futex
  0.43    0.000097          10         9           clone3
  0.37    0.000083           4        19           rt_sigprocmask
  0.27    0.000062           6         9           mprotect
  0.26    0.000058           7         8           mmap
  0.01    0.000002           1         2           recvfrom
  0.00    0.000000           0         6           brk
  0.00    0.000000           0         5           rt_sigaction
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           uname
  0.00    0.000000           0         1           readlink
  0.00    0.000000           0         2         1 arch_prctl
  0.00    0.000000           0         1           epoll_create
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         1           prlimit64
  0.00    0.000000           0         1           getrandom
  0.00    0.000000           0         1           rseq
------ ----------- ----------- --------- --------- ----------------
100.00    0.022704                    86         2 total
```


```ops run --syscall-summary my_app -p 8080 -c config.json```

```time      seconds   usecs/call        calls       errors syscall           
------ ------------ ------------ ------------ ------------ ------------------
 51.29     0.016627          145          115              write             
 35.01     0.011350          183           62            6 openat            
  4.92     0.001595            7          222              clone3            
  3.06     0.000994          497            2              fdatasync         
  0.84     0.000273           14           19              epoll_wait        
  0.63     0.000207            1          183              mprotect          
  0.58     0.000189            5           38              read              
  0.56     0.000182            8           23              pread64           
  0.43     0.000141            2           67            3 futex             
  0.38     0.000124            1          218              exit              
  0.34     0.000111            8           14              getdents64        
  0.26     0.000087            2           37              mmap              
  0.22     0.000073            1           87              uname             
  0.18     0.000061            1           45              close             
  0.18     0.000060           20            3              rename            
  0.10     0.000035           12            3              fsync             
  0.10     0.000033            0          223              set_robust_list   
  0.09     0.000031            0          885              rt_sigprocmask    
  0.07     0.000025            2           14              stat              
  0.06     0.000020            4            5            5 mkdir             
  0.05     0.000018           18            1              setsockopt        
  0.05     0.000017           17            1              socket            
  0.05     0.000017            6            3              unlink            
  0.04     0.000016            2            9            2 epoll_ctl         
  0.04     0.000015            1           17              munmap            
  0.04     0.000015            8            2              accept            
  0.04     0.000014            0         1118              gettid            
  0.04     0.000013            7            2              recvfrom          
  0.03     0.000011            1            9              fstatfs           
  0.03     0.000010            3            3              access            
  0.03     0.000010           10            1              fallocate         
  0.02     0.000009            1           12              newfstatat        
  0.02     0.000007            7            1              readlink          
  0.00     0.000003            3            1              sched_getaffinity 
  0.00     0.000003            3            1              getcwd            
  0.00     0.000003            0          218              madvise           
  0.00     0.000002            0            6              brk               
  0.00     0.000002            2            1              listen            
  0.00     0.000002            0           45            1 fcntl             
  0.00     0.000002            2            1              bind              
  0.00     0.000002            2            1              epoll_create      
  0.00     0.000002            0           18              getpid            
  0.00     0.000001            0            6            6 sync_file_range   
  0.00     0.000001            0            5              rt_sigaction      
  0.00     0.000001            1            1              prlimit64         
  0.00     0.000001            1            1              getrandom         
  0.00     0.000000            0            1            1 syscall_334       
  0.00     0.000000            0            5            5 readahead         
  0.00     0.000000            0            5              fadvise64         
  0.00     0.000000            0            1              fstat             
  0.00     0.000000            0            1              geteuid           
  0.00     0.000000            0            1              getuid            
  0.00     0.000000            0            2            1 arch_prctl        
  0.00     0.000000            0            1              getegid           
  0.00     0.000000            0            1              ftruncate         
  0.00     0.000000            0            1              getgid            
  0.00     0.000000            0            1              set_tid_address   
------ ------------ ------------ ------------ ------------ ------------------
100.00     0.032415                      3769           30 total             
```


Additionally, is there any support for capturing the strace -tt output when running the app using ops run? Here is the output of strace -tt ./my_app.
```
12:50:12.971840 futex(0xf95328, FUTEX_WAKE_PRIVATE, 1) = 0
12:50:12.972227 recvfrom(20, "GET /get HTTP/1.1\r\nHost: lo"..., 1024, 0, NULL, NULL) = 137
12:50:12.972731 mmap(NULL, 8392704, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f041dffc000
12:50:12.972854 mprotect(0x7f041dffd000, 8388608, PROT_READ|PROT_WRITE) = 0
12:50:12.972939 rt_sigprocmask(SIG_BLOCK, ~[], [], 8) = 0
12:50:12.973031 clone3({flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, child_tid=0x7f041e7fc910, parent_tid=0x7f041e7fc910, exit_signal=0, stack=0x7f041dffc000, stack_size=0x7ffdc0, tls=0x7f041e7fc640} => {parent_tid=[3407107]}, 88) = 3407107
12:50:12.973188 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
12:50:12.973246 futex(0xfa0f58, FUTEX_WAKE_PRIVATE, 1) = 1
12:50:12.973296 mmap(NULL, 8392704, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f041d7fb000
12:50:12.973362 mprotect(0x7f041d7fc000, 8388608, PROT_READ|PROT_WRITE) = 0
12:50:12.973417 rt_sigprocmask(SIG_BLOCK, ~[], [], 8) = 0
12:50:12.973481 clone3({flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, child_tid=0x7f041dffb910, parent_tid=0x7f041dffb910, exit_signal=0, stack=0x7f041d7fb000, stack_size=0x7ffdc0, tls=0x7f041dffb640} => {parent_tid=[3407108]}, 88) = 3407108
12:50:12.973578 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
12:50:12.973634 futex(0xfa0f58, FUTEX_WAKE_PRIVATE, 1) = 1
12:50:12.973684 futex(0xf9537c, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY
```

This is my ```ops profile``` output: 
```Ops version: 0.1.42
Nanos version: 0.1.52
Qemu version: 4.2.1
OS: linux
Arch: amd64
Virtualized: false```

**Reply by francescolavra** [reply]
No, there is currently no support for an strace option that outputs timestamps as `-tt` does, although it could be added in the future.
The strace output from your Linux invocation looks quite different from the Nanos output (e.g. there are no openat or write syscalls in the Linux output): when capturing the above logs, did you make requests to the same endpoints (the same number of times and in the same sequence) in the two setups?

**Reply by Roopsai507** [reply]
Yes, in both setups, I made the same single request following with /exit request.

**Reply by francescolavra** [reply]
I understand now why the two logs are different: by default, the Linux strace tool traces syscalls made from the main thread only, but your application spawns different threads, which presumably make other syscalls.
To have strace follow all threads (which is what the Nanos strace klib does), you have to add  the `-f` command line flag, as in `strace -f -c ./my_app`.

**Reply by Roopsai507** [reply]
Yes strace -f -c gave similar output
```
time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 63.81    0.053644         631        85         8 futex
 20.14    0.016935         604        28           epoll_wait
  4.83    0.004063          18       225           clone3
  4.52    0.003799          31       122           write
  4.42    0.003719          16       221           madvise
  0.57    0.000479           0       897           rt_sigprocmask
  0.40    0.000339           0      1118           gettid
  0.18    0.000150         150         1           fallocate
  0.18    0.000150           0       226           rseq
  0.14    0.000117           0       226           set_robust_list
  0.12    0.000100          33         3           rename
  0.12    0.000098           0       185           mmap
  0.11    0.000093           1        64         1 openat
  0.10    0.000083           0       329           mprotect
  0.09    0.000078           0       155           munmap
  0.05    0.000045           0        52           close
  0.04    0.000033           1        18           getpid
  0.04    0.000032           2        11         2 epoll_ctl
  0.03    0.000023           0        42           read
  0.02    0.000019           0        23           pread64
  0.02    0.000015           0        45           fcntl
  0.02    0.000013           0        14           getdents64
  0.01    0.000009           0        87           uname
  0.01    0.000008           0        14           stat
  0.01    0.000006           0        14           newfstatat
  0.00    0.000004           0         5         5 mkdir
  0.00    0.000004           0         5           fadvise64
  0.00    0.000003           3         1           ftruncate
  0.00    0.000003           0         9           fstatfs
  0.00    0.000003           0         5           readahead
  0.00    0.000002           0         3           access
  0.00    0.000002           0         6           sync_file_range
  0.00    0.000001           1         1           fstat
  0.00    0.000001           0         2           accept
  0.00    0.000001           0         2           recvfrom
  0.00    0.000000           0         1           lseek
  0.00    0.000000           0         6           brk
  0.00    0.000000           0         5           rt_sigaction
  0.00    0.000000           0         3           socket
  0.00    0.000000           0         3           bind
  0.00    0.000000           0         3           listen
  0.00    0.000000           0         3           setsockopt
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         3           fsync
  0.00    0.000000           0         2           fdatasync
  0.00    0.000000           0         1           getcwd
  0.00    0.000000           0         3           unlink
  0.00    0.000000           0         1           readlink
  0.00    0.000000           0         2         1 arch_prctl
  0.00    0.000000           0         1           sched_getaffinity
  0.00    0.000000           0         1           epoll_create
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           prlimit64
  0.00    0.000000           0         1           getrandom
------ ----------- ----------- --------- --------- ----------------
100.00    0.084074                  4286        17 total```

**Reply by francescolavra** [reply]
OK, so the differences between the 2 syscall summary outputs don't seem significant enough to explain the large difference in latency when handling the request. To get more insights as to what might cause this latency, you could either:
- timestamp in your C code the beginning and end of the request in your request handler, as well as the intermediate steps, as @eyberg suggested earlier, and compare the timestamp sequences you get when running your app natively and in a Nanos VM
- enable full tracing (with `strace -f` when executing natively, with `--trace` _and_ the strace klib when executing with Ops), and try to spot significant differences in the 2 outputs; since those outputs are likely going to be pretty verbose (especially the Nanos output which will log many internal kernel events), it may take some time for you to make sense of them

One thing to keep in mind when comparing the 2 setups is that the database files being used by your request handler must be in the same state (cached vs non-cached). So if you want to compare things when files are not cached, you have to drop the caches in your Linux host machine before starting the native Linux app and hitting your endpoint; and if you want to compare things when files are cached, you have to make a first "warm-up" request (which will cache the relevant files), and then compare at the second request.
